climada/util/calibrate/func.py

```diff
@@ -0,0 +1,16 @@
+"""Default functions"""
+from typing import Sequence, Optional
+
+import pandas as pd
+import numpy as np
+
+from climada.engine import Impact
+
+def rmse(impact: pd.DataFrame, data: pd.DataFrame):
+    return np.sqrt(np.mean(((impact - data) ** 2).to_numpy()))
+
```

Comment:
I wouldn't use "+ 1" in such a general setting since we do not know the scaling of the impact variable. In some applications it might be an abstract indicator that has a very small value range. I would rather use something like ` + np.spacing(1)`.

Reply:
I agree, but would make it a variable, as for certain cases 1 might be a bad choice.

Reply:
You can make it a variable, but `np.spacing(1)` should be general enough.

Reply:
Aaa yes, I misunderstood what spacing does. I think the only 'problem' might arise if there are negative and positive values, but this is probably to much of an edge case.

Reply:
@tovogt Very good idea with `spacing`! I was not even aware of that function. However, I found that `scikit-learn` is a dependency of Climada and they define error measures useful for regression problems. Maybe we just borrow them? Note that they also define the mean squared logarithmic error with "+1": https://scikit-learn.org/stable/modules/model_evaluation.html#mean-squared-logarithmic-error

@chahank I am unsure about negative values. It is common to compute log(x) - log(y) instead of log(x/y) due to numerical accuracy and the former does not support any negative numbers

Reply:
Any particular feelings towards using the "unrooted" versions of the squared errors? I really have no idea when to use what

Reply:
Taking the square root or not means having a "norm" or a "norm squared". And "squared norms" are mathematically much better behaved in optimization contexts than "norms" because they are strictly convex and the derivative at 0 is continuous. So yes, it's probably better to remove the sqrt.

However, I still think that the `+1` term is not save. Maybe it's numerically more robust if you take `log(x) - log(y)` instead of `log(x/y)` because you avoid subtracting very large numbers. But if you really have an application where your values are between 0 and 1, then this function behaves completely different than in cases where your values are orders of magnitude larger. Close to 0, the mapping `x -> log(1 + x)` is almost the same as `x -> x`...

`MSLE(100, 200)` is approximately equal to `MSLE(1000, 2000)`, but MSLE(0.1, 0.2) is completely different:
```python
>>> import numpy as np
>>> msle = lambda x, y: (np.log(1 + x) - np.log(1 + y))**2
>>> msle(100, 200)
0.4735977563158361
>>> msle(1000, 2000)
0.4797606358194477
>>> msle(10000, 20000)
0.48038370689796994
>>> msle(0.1, 0.2)
0.007570979725631454
>>> msle(0.01, 0.02)
9.706774520097977e-05
```

Reply:
@tovogt Thanks for digging into that! You are correct about the "distortion" effects of the MSLE with values between 0 and 1. However, I would argue that these values rarely occur in Climada impact calculations. Overall, it seems like this is treated as a "property" the user has to be aware of.

Instead of using our own cost functions I now suggest relying on those defined by `scikit.metrics`, see https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics. The metrics there are typically squared and the MSLE is also defined with +1. Not supplying our own cost functions should signal sufficiently to the user that they have to decide for a metric themselves, and be more aware of these intricacies.